File: leetcode/leetcode_cpp/clone-graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

/*
https://leetcode.com/problems/clone-graph

Solution1. DFS
time: o(n)
space: o(n)

Solution2. BFS
time: o(n)
space: o(n)
*/

class Solution {
public:
    Node* dfs(Node* node, unordered_map<Node*, Node*>& hm) {
        if (!node) return nullptr;
        if (hm.count(node)) return hm[node];
        
        hm[node] = new Node(node->val);
        int nCount = node->neighbors.size();
        for (int i=0; i<nCount; ++i) {
            auto n = node->neighbors[i];
            hm[node]->neighbors.push_back(dfs(n, hm));
        }
        return hm[node];
    }

    Node* cloneGraph_1(Node* node) {
        unordered_map<Node*, Node*> hm;
        return dfs(node, hm);        
    }

    Node* cloneGraph_2(Node* node) {
        if (!node) return nullptr;
        queue<pair<Node*, Node*>> q;
        q.push({nullptr, node});

        unordered_map<Node*, Node*> hm;
        Node* root = nullptr;
        while (!q.empty()) {
            int qSize = q.size();
            for (int i=0; i<qSize; ++i) {
                auto nodePair = q.front(); q.pop();
                auto parent = nodePair.first;
                auto child = nodePair.second;
                if (!child) continue;
                if (!hm.count(child))
                {
                    auto clonedChild = new Node(child->val);
                    hm[child] = clonedChild;

                    if (parent)
                        parent->neighbors.push_back(clonedChild);
                    else
                        root = clonedChild;
                    
                    for (auto neighbor : child->neighbors) {
                        q.push({clonedChild, neighbor});
                    }
                }
                else {
                    parent->neighbors.push_back(hm[child]);
                }
            }
        }
        return root;
    }

    Node* cloneGraph(Node* node) {
        return cloneGraph_2(node);        
    }
};

File: leetcode/leetcode_py/clone-graph.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val = 0, neighbors = None):
        self.val = val
        self.neighbors = neighbors if neighbors is not None else []
"""

'''
https://leetcode.com/problems/clone-graph

Solution1. DFS
time: o(n)
space: o(n)

Solution2. BFS
time: o(n)
space: o(n)
'''
from typing import Optional
class Solution:
    def cloneGraph_1(self, node: Optional['Node']) -> Optional['Node']:
        hm = {}
        def dfs(n):
            if not n: return None
            if n in hm: return hm[n]
            cloned = Node(n.val)
            hm[n] = cloned
            for neighbor in n.neighbors:
                cloned.neighbors.append(dfs(neighbor))
            return cloned

        return dfs(node)
    def cloneGraph_2(self, node: Optional['Node']) -> Optional['Node']:
        if not node: return None

        q = []
        q.append([None, node])

        hm = {}
        while q:
            qSize = len(q)
            for i in range(qSize):
                parent, child = q.pop(0)
                if child in hm:
                    parent.neighbors.append(hm[child])
                else:
                    clonedChild = Node(child.val)
                    hm[child] = clonedChild
                    if parent:
                        parent.neighbors.append(clonedChild)

                    for neighbor in child.neighbors:
                        q.append([clonedChild, neighbor])
        return hm[node]

        
    def cloneGraph(self, node: Optional['Node']) -> Optional['Node']:
        return self.cloneGraph_2(node)

        
        